File: built-in plugins dsp/mb-piano-crystal.cpp
/**
 * MB Crystal Piano
 * Category : instrument
 * Type     : piano
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Bright crystalline piano with shimmering highs
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_PIANO_CRYSTAL_H
#define MB_PIANO_CRYSTAL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbPianoCrystal : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-piano-crystal";
    static constexpr const char* PLUGIN_NAME    = "MB Crystal Piano";
    static constexpr const char* PLUGIN_TYPE    = "piano";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float shimmer = 0.6f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbPianoCrystal() = default;
    ~MbPianoCrystal() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.shimmer = std::clamp(params.shimmer, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Crystal Piano
        return input;
    }
};

#endif // MB_PIANO_CRYSTAL_H

File: built-in plugins dsp/mb-piano-crystal.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbPianoCrystalPlugin: PluginDefinition = {
    id: 'mb-piano-crystal', slug: 'mb-piano-crystal', name: 'MB Crystal Piano', category: 'instrument', type: 'piano', version: '1.0.0',
    description: 'Bright crystalline piano with shimmering highs', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'sine', detune: 0, gain: 0.5 }, { type: 'sine', detune: 1200, gain: 0.3 }, { type: 'sine', detune: 2400, gain: 0.2 }],
    envelope: { attack: 0.001, decay: 0.4, sustain: 0.5, release: 0.8 },
    parameters: [
      { id: 'shimmer', name: 'Shimmer', type: 'float', defaultValue: 0.6, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { shimmer: 0.6, volume: 0.8 },
  };

export default MbPianoCrystalPlugin;
